pkg/ruler/ruler.go

```diff
@@ -444,3 +450,131 @@ func (r *Ruler) newManager(ctx context.Context, userID string) (*promRules.Manag
 	}
 	return promRules.NewManager(opts), nil
 }
+
+// GetRules retrieves the running rules from this ruler and all running rulers in the ring if
+// sharding is enabled
+func (r *Ruler) GetRules(ctx context.Context, userID string) ([]*rules.RuleGroupDesc, error) {
+	rgs, err := r.getRules(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.cfg.EnableSharding {
+		rulers, err := r.ring.GetAll()
+		if err != nil {
+			return nil, err
+		}
+
+		ctx, err = user.InjectIntoGRPCRequest(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("unable to inject user ID into grpc request, %v", err)
+		}
+
+		for _, rlr := range rulers.Ingesters {
+			// Don't retrieve rules using GRPC from the same ingester
```

Comment:
Have you actually considered retrieving the rules from the same ingester from gRPC too? I could simplify a bit the code because the entire function could be:

```
if r.cfg.EnableSharding {
   // Load from all ingesters via gRPC
} else {
   return r.getRules(userID)
}
```

Moreover, from a readability perspective, may be more clear if we rename `getRules()` to `getLocalRules()` and this block of code is moved to another function like `getShardedRules()` so that this function code will just be:

```
if r.cfg.EnableSharding {
   return r.getShardedRules(userID)
}

return r.getLocalRules(userID)
```

Reply:
I had considered using GPRC for all of the rules but thought it was strange for the local use case. That snippet will work quite well though